include/hip/hcc_detail/helpers.hpp

```diff
@@ -48,7 +48,21 @@ namespace hip_impl {
 template <typename...>
 using void_t_ = void;
 
-#if (__cplusplus < 201402L)
+#if ((__cplusplus >= 201703L) && defined(__cpp_lib_is_invocable))
+// C++17
+template <FunctionalProcedure F, typename... Ts>
+using is_callable_impl = std::is_invocable<F, Ts...>;
```

Comment:
This wont work. It needs to be:

```cpp
template <FunctionalProcedure F, typename... Ts>
struct is_callable_impl<F(Ts...)> : std::is_invocable<F, Ts...> {};
```

Reply:
This is why unit test should be added.

Reply:
This template alias works since call is being made on type and both classes have it.
Will add a test case to check for it.

Reply:
As a template alias, when instantiating `is_callable_impl<F(Ts...)>` it will instantiate `is_invocable<F(Ts...)>` which is wrong. From [`is_invocable`](https://en.cppreference.com/w/cpp/types/is_invocable) documentation, it cleary shows that it needs to be instantiated as `std::is_invocable<F, Ts...>` not as `is_invocable<F(Ts...)>`.

Reply:
Also, currently, this code is not being used at all either, since `__cpp_lib_is_invocable` won't be defined unless the `<version>` header is included as I mentioned [here](https://github.com/ROCm-Developer-Tools/HIP/pull/1631#discussion_r343236378).